server/src/main/java/org/apache/iotdb/db/query/timegenerator/EngineTimeGenerator.java

```diff
@@ -51,12 +54,17 @@ private void initNode(QueryContext context) throws StorageEngineException {
 
   @Override
   public boolean hasNext() throws IOException {
-    return operatorNode.hasNext();
+    return (cacheTimes != null && cacheTimes.hasMoreData()) || operatorNode.hasNext();
   }
 
   @Override
   public long next() throws IOException {
-    return operatorNode.next();
+    if (cacheTimes == null || !cacheTimes.hasMoreData()) {
```

Comment:
I prefer to implement next() like this:

public BatchData nextBatch() throws IOException {
    if (hasCachedBatchData || hasNextBatch()) {
      hasCachedBatchData = false;
      return batchData;
    }
    throw new IOException("no next batch");
  }

Reply:
this return value was `long` not batch, but  i'm impl with your suggestion